---
title: "match malepools to parent pool"
author: "Casey_Flay"
date: "14/02/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr)
library(RLinuxModules)
library(data.table)
library(glue)
library(here)
library(tidyverse)
library(stringr)
library(ggplot2)
library(magrittr)
library(future)
library(future.batchtools)
plan(batchtools_openlava, resources = list(threads=1))
setDTthreads(1)
module("load openlava asub/2.2 bwa/0.7.12 samtools/1.9 bcftools/1.10.2 vcftools")
```
*pull out russell alleles that are present at a position with a depth count in a 3:1 ratio/ 75%:25% if the alleles in rusell diverge from the 25% ratio in the male pools it should indicate selection on that allele. hopefully in the right spot on chr25.*
*get the alleles that are alts in russell that are not alts in the female parent pools to reduce the numbers*

```{r update testpooldt}
testpooldt <- data.table("SAMPLE"= c("A1", "B2", "C3", "E11", "F12", "G13", "H14", "I15", "J16"))
testpooldt <- testpooldt[, `:=` (with="withhom", chr="_chr25_",  work="pile", vcf=".vcf.gz", filterquery="f", path="/powerplant/workspace/hrtcdf/github/FSTs/malepool/")]
testpooldt

```


```{r run future.batchtools to reduce the time required to run the large files}
files      <-  paste(testpooldt$path, testpooldt$SAMPLE, "_chr25_pile.vcf.gz", sep = "")
#files
dat <- future_lapply(files, function(x) {
  require(glue)
  require(data.table)
  warn(message = x)
  cmd <- glue::glue("bcftools view {x} -Ou | bcftools query  -f'[[%SAMPLE]]\t%POS\t%REF\t%ALT\t%ALT{{0}}\t%ALT{{1}}\t%ALT{{2}}\t%DP\t%AD\t%AD{{0}}\t%AD{{1}}\t%AD{{2}}\t%AD{{3}}\n'")
  warn(message = cmd)
  data.table::fread(cmd=cmd)
  })
  #names(dat) <- testpooldt$SAMPLE
getwd()
```
##col.names = c(paste0(testpooldt$SAMPLE,"SAMPLE"),'POS',paste0(testpooldt$SAMPLE,'REF'), paste0(testpooldt$SAMPLE,'ALT'), paste0(testpooldt$SAMPLE,'ALT0'), paste0(testpooldt$SAMPLE,'ALT1'), paste0(testpooldt$SAMPLE,'ALT2'),paste0(testpooldt$SAMPLE,'ALT3'), paste0(testpooldt$SAMPLE,'DP'),paste0(testpooldt$SAMPLE,'AD0'),paste0(testpooldt$SAMPLE,'AD1'), paste0(testpooldt$SAMPLE,'AD2'), key = 'POS')

```{r male_pool_ofparents, A1, and Russell}
#ALT{0} retrieves the first alternate allele while AD{0} retrieves the depth for the reference. col names adjusted to make ALT1 the same as AD1.

#malepools
testpooldt <- c("A1", "B2", "C3", "E11", "F12", "G13", "H14", "I15", "J16")
if (file.exists("malepools.csv")){file.remove("malepools.csv")}
for (i in testpooldt){
vcfcmd      <- glue::glue("bcftools view '/powerplant/workspace/hrtcdf/github/FSTs/malepool/{i}_chr25_pile.vcf.gz' -Ou | bcftools query  -f'[%SAMPLE]\t%POS\t%REF\t%ALT\t%ALT{{0}}\t%ALT{{1}}\t%ALT{{2}}\t%DP\t%AD\t%AD{{0}}\t%AD{{1}}\t%AD{{2}}\t%AD{{3}}\n'")
testpools   <- fread(cmd=vcfcmd, col.names = c('SAMPLE','POS','REF','ALT','ALT1','ALT2','ALT3','DP','AD','AD0','AD1','AD2','AD3'), key = 'POS')

fwrite(testpools,file="malepools.csv", append=TRUE)
}
testpools   <- fread('/powerplant/workspace/hrtcdf/github/FSTs/malepool/malepools.csv')

##parent pools for males
vcfcmd1      <- "bcftools view '/powerplant/workspace/hrtcdf/github/FSTs/malepool/malepoolparents/pileCK06_01CK20_01CK51_06.vcf' -Ou | bcftools query  -f'[%SAMPLE]\t%POS\t%REF\t%ALT\t%ALT{0}\t%ALT{1}\t%ALT{2}\t%DP\t%AD\t%AD{0}\t%AD{1}\t%AD{2}\t%AD{3}\n'"
male_pool_ofparents <- fread(cmd=vcfcmd1, col.names = c('SAMPLE','POS','REF','ALT','ALT1','ALT2','ALT3','DP','AD','REFD','AD1','AD2','AD3'), key = 'POS')
male_pool_ofparents <- male_pool_ofparents[, SAMPLEmp := "mpparents"]

#Russell and alt
vcfcmd2      <- "bcftools view -r 'chr25' '/powerplant/workspace/hrtcdf/github/FSTs/Russellalt/Russellpile.vcf.gz' -Ou | bcftools query  -f'[%SAMPLE]\t%POS\t%REF\t%ALT\t%ALT{0}\t%ALT{1}\t%ALT{2}\t%DP\t%AD\t%AD{0}\t%AD{1}\t%AD{2}\t%AD{3}\n'"
Ru <- fread(cmd=vcfcmd2, col.names = c('SAMPLE','POS','REF','ALT','ALT1','ALT2','ALT3','DP','AD','REFD','AD1','AD2','AD3'), key = 'POS')

#bind pools and remove unneeded 
testpools  <- rbind(testpools, male_pool_ofparents, Ru)
rm(male_pool_ofparents, Ru)

#pA1 <- merge.data.table(male_pool_ofparents, A1, all=TRUE, by ='POS')
#pA1 <- merge.data.table(Ru, pA1, all=TRUE, by ='POS')
rm(male_pool_ofparents, A1, Ru)
test_subset <- testpools[POS<10000]
if (file.exists("test_subset.csv")){  file.remove("test_subset.csv")}
fwrite(test_subset, file="test_subset.csv")
#rm(test_subset)
##full set
if (file.exists("testpools.csv")){  file.remove("testpools.csv")}
fwrite(pA1, file="testpools.csv")
```

```{r pull out russell alleles that are present at a position with a depth count in a 3:1 ratio/ 75%:25% }
# if the alleles in rusell diverge from the 25% ratio d in the male pools it should indicate selection on that allele. hopefully in the right spot on chr25.
head(testpools)
```


```{r cleanup data tables}
####testpools <- fread("test_subset.csv")
testpools <- fread("test_subset.csv")
#remove sites not in male parents and pool A1 
testpools <- testpools[!is.na(DP)]

head(testpools)
#sapply(testpools,class) #get data class change class to numeric for depth
pA1 <- pA1[, `:=` (AD1a=as.numeric(AD1a), AD2a=as.numeric(AD2a), AD3a=as.numeric(AD3a), AD1mp=as.numeric(AD1mp), AD2mp=as.numeric(AD2mp), AD3mp=as.numeric(AD3mp), AD1r=as.numeric(AD1r), AD2r=as.numeric(AD2r), AD3r=as.numeric(AD3r))]

#clean up low depth snps
testpools <- testpools [AD1<5,  `:=` (AD1=".", ALT1=".")
                      ][AD2<5,  `:=` (AD2=".", ALT2=".")
                      ][AD3<5,  `:=` (AD3=".", ALT3=".")]

#clean low depth samples as they may give false positives. This could be increased for better accuracy.
testpools <- testpools[!AD0<5 & !AD1<5] 

#replace ADrepeats in AD1... with "." where there is only reference alleles.
testpools <- testpools[AD==AD0, `:=`(AD1=".", AD2=".", AD3=".")]
#sapply(testpools,class) #get data class

#testpools <- testpools[is.na(AD1mp), AD1mp := "."][is.na(AD1a), AD1a := "."]
```

**up to here**
```{r note snps which are not the same between Russell and the male parent pool of females mp and calculate which ones are 25% of depth.}

#note snps which are not the same between Russell and the male parent pool of females mp and calculate their AF.
testpools    <-       testpools[ALT1r != ALTmp & ALT1r != ALT1mp & ALT1r != ALT2mp & ALT1r != ALT3mp, `:=` ("AD1r_DPr"=(AD1r/DPr))]

#make a new column for those that are 25% of russell alleles
pA1t    <- pA1[, `:=` ("AD1r_DPr25"=AD1r_DPr>0.23 & AD1r_DPr<0.27)]

#calculate deviance for male test pools at those positions.
pA1t    <- pA1[AD1r_DPr25==TRUE, `:=` ("skew"= AD1a/DPa)]


qplot(pA1t$POS, pA1t$skew)
```


**save this step for the full test. this step is not appropriate for the male pool test.**
```{r note snps to remove which are the same between sample A1 and the male parent pool mp}
pA1t    <-       pA1[ALT1a != ALTmp & ALT1a != ALT1mp & ALT1a != ALT2mp & ALT1a != ALT3mp, "difALT1a":=ALT1a
                    ][ALT2a != ALTmp & ALT2a != ALT1mp & ALT2a != ALT2mp & ALT2a != ALT3mp, "difALT2a":=ALT2a
                    ][ALT3a != ALTmp & ALT3a != ALT1mp & ALT3a != ALT2mp & ALT3a != ALT3mp, "difALT3a":=ALT3a]
```

```{r}
pA1t[is.na(pA1t)] <- "."
pA1t <- pA1t[, "difALTa":=paste(difALT1a,difALT2a,difALT3a)]
qplot(pA1t)


```


