---
title: "testpoolmpile"
author: "Casey Flay"
date: "30/09/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(RLinuxModules)
library(data.table)
library(glue)
library(here)
library(tidyverse)
library(stringr)
module("load openlava asub bwa/0.7.12 samtools/1.9 bcftools/1.10.2")
```

```{r}
mainDir <- "/powerplant/workspace/hrtcdf/FSTs/"
subDir <- "bamlink"
royDir <- "/workspace/hrards/projects/hrtcdf/10.alignments/bwa"
outDir <- str_c(mainDir, subDir, sep = "", collapse = TRUE)
outDir

if (dir.exists(mainDir)){
    setwd(file.path(mainDir, subDir))
} else {
    dir.create(file.path(mainDir, subDir))
    setwd(file.path(mainDir, subDir))
}

tpoolfiles <- dir(royDir, pattern = "^HL55WDRXX", full.names = TRUE, include.dirs = TRUE)

file.symlink (tpoolfiles, outDir)
## cant seem to get the full pathname into the symlink file. it doesnt need it.
```


```{r transform filenames to sample and directory for data.table}
files <- dir("/workspace/hrards/projects/hrtcdf/10.alignments/bwa", pattern = "^HL55WDRXX", full.names = TRUE)
apfiles <- str_replace(files, "HL55WDRXX_" , "HL55WDRXX_&")

dt <- {data.table(apfiles) %>%
  separate(apfiles, into = c("dir", "sample"), sep = "&", fill = "left") %>%
  mutate("version" = str_extract(dir, regex("/")))
}

colnames(dt) <- c("dir", "sample", "version")
write.table(dt, 'testpool.tsv')
```


```{r transform filenames to sample, lane and suffix for data.table from link}
#setwd("/powerplant/workspace/hrtcdf/FSTs/TestPool/bamlinkindexed")
wd <- ("/powerplant/workspace/hrtcdf/FSTs/TestPool/bamlinkindexed/")
wd
pattern <- c(".bai", "log", ".tsv")
f <- grep(paste(pattern, collapse = "|"), list.files(path="."),  invert=TRUE, value=TRUE)
f
str1 <- str_replace(f, "XX_" , "XX_&")
str2 <- str_replace(str1, "_L" , "&_L")
str3 <- str_replace(str2, ".bam" , "&.bam&")
str3
lindts <- {data.table(str3) %>%
  separate(str3, into = c("HL55", "sample", "lane", "bam", "version"), sep = "&", fill = "left") %>%
  mutate("version" = str_extract(sample, regex("/")))
}
lindts

colnames(lindts) <- c("HL55", "sample", "lane", "bam", "version" )
#write.table(lindts, 'linkpool.tsv')
```

#Still need to make read groups so each sample lane can be merged into one before mpileup? this will have to be done for the full pool data.
```{r run-mileup for chr25}
submit_mpileup <- function(reference = "*.fa.gz", output = c(), bamfiles = c(), job.init = "asub", job.opts = "-q normal") {
  cmd <- "bcftools mpileup -B -Q 0 -f {reference} -r chr25 -a {mpileuptags} {bamfiles} -Ou | bcftools call -f {bcfcalltags} -p {pvalThres} -m -A -Oz -o {output}"
  tmpf <- tempfile(pattern = "submit_mileup.", fileext = ".sh")
  writeLines(text = c(glue(cmd)), con = tmpf)
  submit <- "{job.init} {job.opts} {tmpf}"
  system(glue(submit), intern = TRUE)
}
mpileuptags <- "DP,AD,ADF,ADR,SP,INFO/AD,INFO/ADF,INFO/ADR"
bcfcalltags <- "GQ,GP"
  pvalThres <- 0.99

bamfiles <- paste(lindts$HL55, lindts$sample, lindts$lane, lindts$bam, sep = "")
#bamfiles
sample   <- paste(lindts$sample,lindts$lane, sep = "")
sample   <- paste(tools::file_path_sans_ext(sample), "25a_bam.gz", sep = ".")
#sample

reference <- dir("/workspace/hrards/projects/hrtcdf/genome/fasta", pattern = ".*\\.gz$", full.names = TRUE)
#reference

#file.exists(bamfiles)
#file.exists(reference)

submit_mpileup(reference = reference, output = sample, bamfiles = bamfiles)
```

## run for unzipped file once working and run through Tassel 5 for hapmap format.
## -e 'F_MISSING>0.0' takes out all individuals with a missing genotype.
```{r, engine='bash' eval=FALSE}

t   <- 8
x   <- f
bwa <- bwa ## change from [bwa_subset] to [bwa] to run full set not including square brackets.
logDir <- "/powerplant/workspace/hrtcdf/GBSremap/15_FilterVCF/logDir"
tts <- logDir bwa
tts

bsub -n t -o logDir/bwa_{x}.log -e $logDir/${bwa}_${x}.err "bcftools view -i 'INFO/MAF[0]>0.05 && AVG(INFO/DP)>16 && QUAL>=20' $sample | bcftools filter -e 'F_MISSING>0.0' -g9 -G9 -o ${bwa}_${x}.vcf && bcftools stats ${bwa}_${x}.vcf > ${bwa}_${x}.vcf.stats && run_pipeline.pl -fork1 -vcf ${bwa}_${x}.vcf -export ${bwa}_${x} -exportType Hapmap -runfork1"
```